### PR TITLE
bpo-12159: document sys.maxsize limit in len() function reference

### PR DESCRIPTION
**Pull request by Zac-HD**
See: 

- BPO-12159:  "needs to be documented",
- BPO-15718:  added note to datamodel docs for `__len__` method but not `len()`, and 
- BPO-21444:  change request / wontfix

https://bugs.python.org/issue12159

### CODE REVIEW

**Comment by cool-RR** [comment]
I think that this is a good addition to the documentation, but the example with `range` is weird. Maybe you mean that big ranges, such as `range(10 ** 100)` will be too big. That isn't clear from your text.

File: Doc/library/functions.rst
Comment (vstinner) [inline Doc/library/functions.rst]:
```suggestion
   on lengths larger than :data:`sys.maxsize`.
```

I don't understand the "such as range()" part. len(range(5)) returns 5, it doesn't fail with overflow?

Comment (Zac-HD) [reply Doc/library/functions.rst]: I've clarified that this is for cases such as `range(2 ** 64)`.

File: Doc/library/functions.rst
Comment (vstinner) [inline Doc/library/functions.rst]: I suggest to use a number even larger, since some developers may expect maxsize to be equal to 2**64.

```suggestion
   :class:`range(2 ** 100) <range>`.
```

With `2**100`, there is a lower risk of confusion ;-)

**Review by vstinner — APPROVED** [review]
LGTM, but I would prefer to get a second review. @csabella @serhiy-storchaka @JulienPalard: What do you think of this len note?

**Review by csabella — APPROVED** [review]
LGTM.  Since this has been brought up multiple times, I think it's best to document it, especially since ``__len__`` had a note added in the past.  Thanks for the patch, @Zac-HD!

File: Doc/library/functions.rst
Comment (csabella) [inline Doc/library/functions.rst]: There is a directive for implementation details:
```
.. impl-detail::
```

**Review by csabella — CHANGES_REQUESTED** [review]

File: Doc/library/functions.rst
Comment (Zac-HD) [reply Doc/library/functions.rst]: Thanks @csabella - done :smile:

**Comment by Zac-HD** [comment]
I have made the requested changes; please review again

**Pull request merged**
